typing/shape.ml

```diff
@@ -209,73 +206,251 @@ let proj ?uid t item =
   | _ ->
       { uid; desc = Proj (t, item) }
 
-let rec app ?uid f ~arg =
-  match f.desc with
-  | Abs (var, body) ->
-      let res = subst var ~arg body in
-      overwrite_uid uid res
-  | _ ->
+let app ?uid f ~arg =
       { uid; desc = App (f, arg) }
 
-and subst var ~arg t =
-  match t.desc with
-  | Var id when var = id -> arg
-  | Abs (v, e) ->
-      abs ?uid:t.uid v (subst var ~arg e)
-  | App (f, e) ->
-      app ?uid:t.uid (subst var ~arg f) ~arg:(subst var ~arg e)
-  | Struct m ->
-      { t with desc = Struct (Item.Map.map (fun s -> subst var ~arg s) m) }
-  | Proj (t, item) ->
-      proj ?uid:t.uid (subst var ~arg t) item
-  | Comp_unit _ | Leaf | Var _ ->
-      t
-
-
 module Make_reduce(Params : sig
   type env
   val fuel : int
   val read_unit_shape : unit_name:string -> t option
   val find_shape : env -> Ident.t -> t
 end) = struct
-  let fuel = ref (-1)
-  let rec reduce env t =
-    if !fuel < 0 then
-      t
+  module IdentMap = Map.Make(Ident)
```

Comment:
This already exists as `Ident.Map`.

Reply:
https://github.com/ocaml/ocaml/blob/trunk/typing/ident.mli#L20 , you can grep your codebase from the past, it's already used.

Reply:
To my defense, it's hard to tell what names are defined in an ML module...